yoyodyne/train.py

```diff
@@ -214,103 +68,352 @@ def main(
+        arch, attention, train_set.config.has_features
     )
-    # Writes development set predictions using the best checkpoint,
-    # if a predictions path is specified.
-    # TODO: Add beam-width option so we can make predictions with beam search.
-    if dev_predictions:
-        best_model = model_cls.load_from_checkpoint(
-            ckp_callback.best_model_path
-        )
-        predict.write_predictions(
-            best_model,
-            dev_loader,
-            dev_predictions,
-            arch,
-            batch_size * 2,
-            config,
+    expert = (
+        models.expert.get_expert(
+            train_set,
+            epochs=oracle_em_epochs,
+            oracle_factor=oracle_factor,
+            sed_params_path=sed_params,
         )
+        if arch in ["transducer"]
+        else None
+    )
+    # Please pass all arguments by keyword and keep in lexicographic order.
+    return model_cls(
+        arch=arch,
+        attention_heads=attention_heads,
+        beta1=beta1,
+        beta2=beta2,
+        bidirectional=bidirectional,
+        decoder_layers=decoder_layers,
+        dropout=dropout,
+        embedding_size=embedding_size,
+        encoder_layers=encoder_layers,
+        end_idx=train_set.end_idx,
+        expert=expert,
+        features_vocab_size=getattr(train_set, "features_vocab_size", -1),
+        features_idx=getattr(train_set, "features_idx", -1),
+        hidden_size=hidden_size,
+        learning_rate=learning_rate,
+        max_decode_length=max_decode_length,
+        max_sequence_length=max_sequence_length,
+        optimizer=optimizer,
+        output_size=train_set.target_vocab_size,
+        pad_idx=train_set.pad_idx,
+        scheduler=scheduler,
+        start_idx=train_set.start_idx,
+        train_set=train_set,
+        vocab_size=train_set.source_vocab_size,
+        warmup_steps=warmup_steps,
+    )
+
+
+def train(
+    trainer: pl.Trainer,
+    model: pl.LightningModule,
+    train_loader: data.DataLoader,
+    dev_loader: data.DataLoader,
+    train_from: Optional[str] = None,
+) -> str:
+    """Trains the model.
+
+    Args:
+         trainer (pl.Trainer).
+         model (pl.LightningModule).
+         train_loader (data.DataLoader).
+         dev_loader (data.DataLoader).
+         train_from (str, optional): if specified, starts training from this
+            checkpoint.
+
+    Returns:
+        (str) Path to best checkpoint.
+    """
+    trainer.fit(model, train_loader, dev_loader, ckpt_path=train_from)
+    ckp_callback = trainer.callbacks[-1]
+    # TODO: feels flimsy.
+    assert type(ckp_callback) is callbacks.ModelCheckpoint
+    return ckp_callback.best_model_path
+
+
+def main() -> None:
+    """Trainer."""
+    parser = argparse.ArgumentParser(description=__doc__)
+    parser.add_argument(
+        "--experiment", required=True, help="Name of experiment"
+    )
+    # Path arguments.
+    parser.add_argument(
+        "--train",
```

Comment:
I feel the same about this as `--predict` in the other file. But as long as they are consistent I guess it seems fine!